ovn-tester/ovn_utils.py

```diff
@@ -35,15 +119,26 @@ def add_port(self, port, bridge, internal=True, ifaceid=None):
     def del_port(self, port):
         self.run(f'del-port {port.name}')
 
+    def create_vm_port_addr_str(self, name, addr, plen, gw):
+        cmd = f'ip -n {name} addr add {addr}/{plen} dev {name} ; ' \
+              f'ip -n {name} route add default via {gw} ; '
+        return cmd
+
     def bind_vm_port(self, lport):
         cmd = f'bash -c \'ip netns add {lport.name} ; ' \
+              f'ip netns exec {lport.name} sysctl -w ' \
+              f'net.ipv6.conf.all.optimistic_dad=1 ; ' \
               f'ip link set {lport.name} netns {lport.name} ; ' \
               f'ip -n {lport.name} link set {lport.name} ' \
               f'address {lport.mac} ; ' \
-              f'ip -n {lport.name} addr add {lport.ip}/{lport.plen} ' \
-              f'dev {lport.name} ; ' \
-              f'ip -n {lport.name} link set {lport.name} up ; ' \
-              f'ip -n {lport.name} route add default via {lport.gw}\''
+              f'ip -n {lport.name} link set {lport.name} up ; '
+        if lport.ip:
+            cmd += self.create_vm_port_addr_str(lport.name, lport.ip,
+                                                lport.plen, lport.gw)
+        if lport.ip6:
+            cmd += self.create_vm_port_addr_str(lport.name, lport.ip6,
+                                                lport.plen6, lport.gw6)
+        cmd += '\''
```

Comment:
To avoid this complex "quote management" I'd change the function to (I also fixed up some unneeded extra whitespace):
```
    def bind_vm_port(self, lport):
        cmd = f'ip netns add {lport.name}; ' \
              f'ip netns exec {lport.name} sysctl -w ' \
              f'net.ipv6.conf.all.optimistic_dad=1; ' \
              f'ip link set {lport.name} netns {lport.name}; ' \
              f'ip -n {lport.name} link set {lport.name} ' \
              f'address {lport.mac}; ' \
              f'ip -n {lport.name} link set {lport.name} up; '
        if lport.ip:
            cmd += self.create_vm_port_addr_str(lport.name, lport.ip,
                                                lport.plen, lport.gw)
        if lport.ip6:
            cmd += self.create_vm_port_addr_str(lport.name, lport.ip6,
                                                lport.plen6, lport.gw6)
        # Run as a single invocation:
        cmd = f'bash -c \'{cmd}\''
        self.run(cmd, prefix="")
```

Reply:
@dceara ack, I will fix it